edge-modules/DirectMethodSender/Program.cs

```diff
@@ -24,21 +26,47 @@ public static async Task<int> MainAsync()
             (CancellationTokenSource cts, ManualResetEventSlim completed, Option<object> handler) = ShutdownHandler.Init(TimeSpan.FromSeconds(5), Logger);
             DirectMethodSenderBase directMethodClient = null;
             ModuleClient reportClient = null;
+            Option<Uri> analyzerUrl = Settings.Current.AnalyzerUrl;
+            Option<Uri> testReportCoordinatorUrl = Settings.Current.TestResultCoordinatorUrl;
+
             try
             {
+                Guid batchId = Guid.NewGuid();
+                Logger.LogInformation($"Batch Id={batchId}");
+
                 directMethodClient = await CreateClientAsync(Settings.Current.InvocationSource);
 
-                Option<Uri> analyzerUrl = Settings.Current.AnalyzerUrl;
                 reportClient = await ModuleUtil.CreateModuleClientAsync(
                     Settings.Current.TransportType,
                     ModuleUtil.DefaultTimeoutErrorDetectionStrategy,
                     ModuleUtil.DefaultTransientRetryStrategy,
                     Logger);
-                while (!cts.Token.IsCancellationRequested)
+
+                Logger.LogInformation($"Load gen delay start for {Settings.Current.TestStartDelay}.");
+                await Task.Delay(Settings.Current.TestStartDelay);
+
+                DateTime testStartAt = DateTime.UtcNow;
+                while (!cts.Token.IsCancellationRequested && IsTestTimeUp(testStartAt))
                 {
-                    HttpStatusCode result = await directMethodClient.InvokeDirectMethodAsync(cts);
+                    (HttpStatusCode result, long dmCounter) = await directMethodClient.InvokeDirectMethodAsync(cts);
 
                     // TODO: Create an abstract class to handle the reporting client generation
+                    await testReportCoordinatorUrl.ForEachAsync(
```

Comment:
Can you please add an interface with SendReportAsync (like in twinTester changes) and have 2 implementations, one for trc, one for analyzer. Based on analyzerUrl or trc url you can create an instance of the analyzerReportHandler or trcReportHandler

Reply:
Yeah, I had some idea on how to do this with a nice way by using abstract class: https://msazure.visualstudio.com/One/_workitems/edit/5949787. 
Since it can be used across multiple module, I'm thinking I will do it in Module Utils and then back port it to DM and Twin testers.  Would it be alright with you? I'm thinking of just merging the baseline (function block) of the DM tester first. Then slowly introduce more feature to it.